reconcile/terraform_cloudflare_dns.py

```diff
@@ -176,6 +176,16 @@ def get_desired_state_shard_config(self) -> DesiredStateShardConfig:
         )
 
 
+def get_cloudflare_provider_rps(zone_size: int) -> int:
+    if zone_size <= 50:
+        rps = 2
+    elif zone_size <= 1000:
+        rps = 4
```

Comment:
The 4 r/s is the default.

I wonder if it only made sense to limit for when a zone has over 1000 resource records?

Hopefully, Terraform parallelism (or an attempt at when resources are independent) does not stand in the way here. I suppose the entire provider has a single client.

Reply:
Hm I'm not understanding your question. Do you mean for records <= 1000 so no setting of rps = 2?

Reply:
FYI. We have the 5000 per 5 minutes rate limit for that account which is about 16 rps so from 4 to 6 at provider level should be having a performance relief for this.

Reply:
The reason we are doing this is because the amount of records in a zone increase the amount of request making to Cloudflare provider APIs, therefore to complete a build. This affect not only when people change that large zone, it also affect pr-check when promoting qontract-reconcile for example. And previous we were seeing around 250 seconds for a 1000 records zone, and that was where we drew the line. But there was a strong request for us to manage a zone with up to 1300 records now, we are trying to accommodate to that while not sacrifice the performance too much. I.e not to let a integration run more than 250 seconds. And that's how we decided on 1500 records upper limit with 6 rps. 

>  if we have an issue with 4 r/s when a zone has 50 RRs, then wouldn't a limit of 4 be an issue for a zone with 999 RRs?
So the problem really wasn't that 50 records were taking too long, it was 1300 will take a long time. And 2 rps for 50 records was just to save our throttle quota and those small zone (which will probably be most of them )are still gonna finish within a short amount time.

> Also, if there are no RRs or a zone has over 1000 RRs, then we set the limit to 6 r/s, but why? Does Cloudflare throttle all of the API requests or just for specific endpoints? So, for example, anything else but zone-related operations can achieve more r/s (or have higher parallelism).
We will have zone with more than 1000 records and reasons are like what i mentioned. I was under the impression that this rate limit is global based on https://developers.cloudflare.com/fundamentals/api/reference/limits/ but I don't have a perfect answer to how specifically Cloudflare throttle API requests. 

> With the default of 4 for everything, are we having problems with extensive zones only? That would be my uneducated assumption.
Hope the background info I gave above answers this.

Reply:
[...]
> Hope the background info I gave above answers this.

I appreciate the background and additional details. The "how" was obvious (per the code), but the "why" wasn't.

Thank you!